modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionModificationHandler.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.ignite.internal.replicator.ReplicaResult;
+import org.apache.ignite.internal.replicator.ReplicationGroupId;
+import org.apache.ignite.internal.replicator.message.GetEstimatedSizeWithModifyTsRequest;
+import org.apache.ignite.internal.replicator.message.ReplicaMessagesFactory;
+import org.apache.ignite.internal.replicator.message.ReplicationGroupIdMessage;
+import org.jetbrains.annotations.Nullable;
+
+/** Partition modifications requests handler. */
+class PartitionModificationHandler {
+    private final MessagingService messagingService;
+
+    private static final PartitionReplicationMessagesFactory PARTITION_REPLICATION_MESSAGES_FACTORY =
+            new PartitionReplicationMessagesFactory();
+
+    /** Replica messages factory. */
+    private static final ReplicaMessagesFactory REPLICA_MESSAGES_FACTORY = new ReplicaMessagesFactory();
+
+    private final Function<ReplicationGroupId, CompletableFuture<Replica>> replicaGetter;
+    private final Supplier<UUID> localNode;
+
+    PartitionModificationHandler(
+            MessagingService messagingService,
+            Function<ReplicationGroupId, CompletableFuture<Replica>> replicaGetter,
+            Supplier<UUID> localNode
+    ) {
+        this.messagingService = messagingService;
+        this.replicaGetter = replicaGetter;
+        this.localNode = localNode;
+    }
+
+    /**
+     * Starts the component.
+     */
+    void start() {
+        messagingService.addMessageHandler(PartitionReplicationMessageGroup.class, this::handleMessage);
+    }
+
+    private void handleMessage(NetworkMessage message, InternalClusterNode sender, @Nullable Long correlationId) {
+        if (message instanceof GetEstimatedSizeWithLastModifiedTsRequest) {
+            handleRequestCounter((GetEstimatedSizeWithLastModifiedTsRequest) message, sender, correlationId);
+        }
+    }
+
+    private void handleRequestCounter(
+            GetEstimatedSizeWithLastModifiedTsRequest message,
+            InternalClusterNode sender,
+            Long correlationId
+    ) {
+        assert correlationId != null;
+
+        List<CompletableFuture<ReplicaResult>> replicaReqFutures = new ArrayList<>();
+
+        for (ReplicationGroupIdMessage repl : message.replicas()) {
+            ReplicationGroupId replGrpId = repl.asReplicationGroupId();
+
+            CompletableFuture<Replica> replica = replicaGetter.apply(replGrpId);
+
+            if (replica != null) {
+                replicaReqFutures.add(
+                        replica.thenCompose(r -> {
+                            GetEstimatedSizeWithModifyTsRequest req = REPLICA_MESSAGES_FACTORY.getEstimatedSizeWithModifyTsRequest()
+                                    .groupId(repl)
+                                    .tableId(message.tableId())
+                                    .build();
+
+                            return r.processRequest(req, localNode.get());
+                        }).exceptionally(e -> null)
```

Comment:
Can we return fake value like (0, HybridTimeStamp.MIN_VALUE)
to avoid null checks further?